### PR TITLE
Feature/browser tabs screen

### PR DESCRIPTION
**Pull request by antonijzelinskij**
#8696nvaf0

### CODE REVIEW

File: app/src/main/java/io/novafoundation/nova/app/root/navigation/buy/BuyNavigator.kt
Comment (antonijzelinskij) [inline app/src/main/java/io/novafoundation/nova/app/root/navigation/buy/BuyNavigator.kt:1]: Moved navigators to `app/src/main/java/io/novafoundation/nova/app/root/navigation/navigator` directory

File: app/src/main/java/io/novafoundation/nova/app/root/navigation/navigators/dApp/DAppNavigator.kt [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+import io.novafoundation.nova.app.root.navigation.navigators.BaseNavigator
+import io.novafoundation.nova.feature_dapp_api.DAppRouter
+import io.novafoundation.nova.feature_dapp_impl.presentation.addToFavourites.AddToFavouritesFragment
+import io.novafoundation.nova.feature_dapp_api.presentation.addToFavorites.AddToFavouritesPayload
+import io.novafoundation.nova.feature_dapp_impl.presentation.browser.main.DAppBrowserFragment
+import io.novafoundation.nova.feature_dapp_impl.presentation.search.DappSearchFragment
+import io.novafoundation.nova.feature_dapp_impl.presentation.search.SearchPayload
+
+class DAppNavigator(
+    private val rootNavigationHolder: RootNavigationHolder,
```
Comment (antonijzelinskij) [inline app/src/main/java/io/novafoundation/nova/app/root/navigation/navigators/dApp/DAppNavigator.kt:14]: Use root navigation holder to open dapp fragments on top of the split screen

File: app/src/main/java/io/novafoundation/nova/app/root/navigation/navigators/dApp/DAppNavigator.kt [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+) : BaseNavigator(rootNavigationHolder), DAppRouter {
+
+    override fun openChangeAccount() = performNavigation(R.id.action_open_switch_wallet)
+
+    override fun openDAppBrowser(initialUrl: String) {
+        // Close deapp browser if it is already opened
+        // TODO it's better to provide new url to existing browser
```
Comment (valentunn) [inline app/src/main/java/io/novafoundation/nova/app/root/navigation/navigators/dApp/DAppNavigator.kt:21]: Pending todo

File: feature-dapp-impl/src/main/java/io/novafoundation/nova/feature_dapp_impl/di/DAppFeatureHolder.kt [hunk 3]
```diff
@@ -6,7 +6,7 @@ import io.novafoundation.nova.common.di.scope.ApplicationScope
 import io.novafoundation.nova.core_db.di.DbApi
 import io.novafoundation.nova.feature_account_api.di.AccountFeatureApi
 import io.novafoundation.nova.feature_currency_api.di.CurrencyFeatureApi
-import io.novafoundation.nova.feature_dapp_impl.DAppRouter
+import io.novafoundation.nova.feature_dapp_api.DAppRouter
```
Comment (valentunn) [inline feature-dapp-impl/src/main/java/io/novafoundation/nova/feature_dapp_impl/di/DAppFeatureHolder.kt:9]: I dont think Router should be in `api` module
If some other feature needs some dapp-related navigation, it should declare it in its own `Router` so we dont mix multiple routers in the same feature, which creates more coupled connections between components

File: feature-dapp-impl/src/main/java/io/novafoundation/nova/feature_dapp_impl/presentation/search/DAppSearchCommunicator.kt [hunk 4]
```diff
@@ -12,6 +12,15 @@ interface DAppSearchResponder : InterScreenResponder<SearchPayload, Response>
 
 interface DAppSearchCommunicator : DAppSearchRequester, DAppSearchResponder {
 
-    @Parcelize
-    class Response(val newUrl: String?) : Parcelable
+    sealed interface Response : Parcelable {
+
+        @Parcelize
+        object TabChanged : Response
```
Comment (valentunn) [inline feature-dapp-impl/src/main/java/io/novafoundation/nova/feature_dapp_impl/presentation/search/DAppSearchCommunicator.kt:18]: I dont see this is handled anywhere - do we even need it?

File: feature-dapp-impl/src/main/java/io/novafoundation/nova/feature_dapp_impl/presentation/tab/BrowserTabsViewModel.kt [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+    val tabsFlow = browserTabPoolService.tabStateFlow
+        .map { it.tabs }
+        .mapList {
+            BrowserTabRvItem(
+                tabId = it.id,
+                tabName = it.pageSnapshot.pageName,
+                tabFaviconPath = it.pageSnapshot.pageIconPath,
+                tabScreenshotPath = it.pageSnapshot.pagePicturePath
+            )
+        }
```
Comment (valentunn) [inline feature-dapp-impl/src/main/java/io/novafoundation/nova/feature_dapp_impl/presentation/tab/BrowserTabsViewModel.kt:33]: `shareInBackground`

File: feature-dapp-impl/src/main/java/io/novafoundation/nova/feature_dapp_impl/utils/tabs/models/TabsState.kt [hunk 6]
```diff
@@ -0,0 +1,11 @@
+package io.novafoundation.nova.feature_dapp_impl.utils.tabs.models
+
+class TabsState(
+    val tabs: List<BrowserTab>,
+    val selectedTab: CurrentTabState
+) {
+
+    fun stateId(): String {
+        return tabs.joinToString { it.id } + "_" + selectedTab.stateId()
+    }
```
Comment (valentunn) [inline feature-dapp-impl/src/main/java/io/novafoundation/nova/feature_dapp_impl/utils/tabs/models/TabsState.kt:10]: This really looks like we are creating an equivalent to `equals`

File: app/src/main/java/io/novafoundation/nova/app/root/navigation/navigators/dApp/DAppNavigator.kt [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+
+    override fun openAuthorizedDApps() = performNavigation(
+        actionId = R.id.action_mainFragment_to_authorizedDAppsFragment
+    )
+
+    override fun openTabs() = performNavigation(
+        actionId = R.id.action_DAppBrowserFragment_to_browserTabsFragment
+    )
+
+    override fun finishTabs() = performNavigation(
```
Comment (valentunn) [inline app/src/main/java/io/novafoundation/nova/app/root/navigation/navigators/dApp/DAppNavigator.kt:57]: `finishTabs` sounds unclear to me.  Maybe we can rename it to something more informative

**Review by valentunn — COMMENTED** [review]
Am I right that we now have the following navigation?

`RootActivity -> SplitFragment -> MainFragment + CollapsedDappView -> BalanceListFragment/StakingDashboardFragment`

And when we move from MainFragment to some secondary flow screen we will have

`RootActivity -> SplitFragment -> SecondaryFlowFragment + CollapsedDappView`

Can you provide some screenshots with dummy `CollapsedDappView` on how it looks like?

File: feature-dapp-impl/src/main/java/io/novafoundation/nova/feature_dapp_impl/presentation/search/DAppSearchCommunicator.kt (hunk 4 above)
Comment (antonijzelinskij) [reply feature-dapp-impl/src/main/java/io/novafoundation/nova/feature_dapp_impl/presentation/search/DAppSearchCommunicator.kt:18]: We need to handle only new url, but I anyway send this response

**Pull request merged**